engine/src/main/java/org/pentaho/di/resource/ResourceUtil.java

```diff
@@ -0,0 +1,186 @@
+   *          The name of the file for the XML to inject in the ZIP archive (optional, can be null)
+   * @return The full VFS filename reference to the serialized export interface XML file in the ZIP archive.
+   * @throws KettleException
+   *           in case anything goes wrong during serialization
+   */
+  public static final TopLevelResource serializeResourceExportInterface( String zipFilename,
+    ResourceExportInterface resourceExportInterface, VariableSpace space, Repository repository,
+    IMetaStore metaStore, String injectXML, String injectFilename ) throws KettleException {
+
+    ZipOutputStream out = null;
+
+    try {
+      Map<String, ResourceDefinition> definitions = new HashMap<String, ResourceDefinition>();
+
+      // In case we want to add an extra pay-load to the exported ZIP file...
+      //
+      if ( injectXML != null ) {
+        ResourceDefinition resourceDefinition = new ResourceDefinition( injectFilename, injectXML );
+        definitions.put( injectFilename, resourceDefinition );
+      }
+
+      ResourceNamingInterface namingInterface = new SequenceResourceNaming();
+
+      String topLevelResource =
+        resourceExportInterface.exportResources( space, definitions, namingInterface, repository, metaStore );
+
+      if ( topLevelResource != null && !definitions.isEmpty() ) {
+
+        // Create the ZIP file...
+        //
+        FileObject fileObject = KettleVFS.getFileObject( zipFilename, space );
+
+        // Store the XML in the definitions in a ZIP file...
+        //
+        out = new ZipOutputStream( KettleVFS.getOutputStream( fileObject, false ) );
+
+        for ( Map.Entry<String, ResourceDefinition> entry : definitions.entrySet() ) {
+          String filename = entry.getKey();
+          ResourceDefinition resourceDefinition = entry.getValue();
+
+          ZipEntry zipEntry = new ZipEntry( resourceDefinition.getFilename() );
+
+          String comment =
+            BaseMessages.getString(
+              PKG, "ResourceUtil.SerializeResourceExportInterface.ZipEntryComment.OriginatingFile", filename,
+              Const.NVL( resourceDefinition.getOrigin(), "-" ) );
+          zipEntry.setComment( comment );
+          out.putNextEntry( zipEntry );
+
+          out.write( resourceDefinition.getContent().getBytes() );
+          out.closeEntry();
+        }
+        String zipURL = fileObject.getName().toString();
+        return new TopLevelResource( topLevelResource, zipURL, "zip:" + zipURL + "!" + topLevelResource );
+      } else {
+        throw new KettleException( BaseMessages.getString( PKG, "ResourceUtil.Exception.NoResourcesFoundToExport" ) );
+      }
+    } catch ( Exception e ) {
+      throw new KettleException( BaseMessages.getString(
+        PKG, "ResourceUtil.Exception.ErrorSerializingExportInterface", resourceExportInterface.toString() ), e );
+    } finally {
+      if ( out != null ) {
+        try {
+          out.close();
+        } catch ( IOException e ) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)
Similar issue at line numbers 149.